File: packages/java/src/main/java/com/ailabtools/sdk/exception/ApiException.java
package com.ailabtools.sdk.exception;

import com.ailabtools.sdk.response.ErrorDetail;
import java.util.Collections;
import java.util.Map;

/** An HTTP or business error returned by the AILabTools API. */
public final class ApiException extends RuntimeException {
    private final Integer statusCode;
    private final String requestId;
    private final String logId;
    private final ErrorDetail detail;
    private final Map<String, Object> raw;

    public ApiException(
            String message,
            Integer statusCode,
            String requestId,
            String logId,
            ErrorDetail detail,
            Map<String, Object> raw) {
        super(message);
        this.statusCode = statusCode;
        this.requestId = requestId;
        this.logId = logId;
        this.detail = detail;
        this.raw = raw == null ? Collections.emptyMap() : Collections.unmodifiableMap(raw);
    }

    public Integer getStatusCode() { return statusCode; }
    public String getRequestId() { return requestId; }
    public String getLogId() { return logId; }
    public ErrorDetail getDetail() { return detail; }
    public Map<String, Object> getRaw() { return raw; }
}

File: packages/php/src/Exception/ApiException.php
<?php

declare(strict_types=1);

namespace AILabTools\Exception;

use AILabTools\Response\ErrorDetail;
use RuntimeException;

final class ApiException extends RuntimeException
{
    /** @param array<string, mixed>|string|null $raw */
    public function __construct(
        string $message,
        public readonly ?int $statusCode = null,
        public readonly ?string $requestId = null,
        public readonly ?string $logId = null,
        public readonly ?ErrorDetail $detail = null,
        public readonly array|string|null $raw = null,
    ) {
        parent::__construct($message);
    }
}
